pkg/debuginfo/manager.go

```diff
@@ -97,47 +104,56 @@ func (di *Manager) removeAsUploading(buildID string) {
 // EnsureUploaded ensures that the extracted or the found debuginfo for the given buildID is uploaded.
 func (di *Manager) EnsureUploaded(ctx context.Context, objFiles []*objectfile.MappedObjectFile) {
 	for _, objFile := range objFiles {
-		buildID := objFile.BuildID
-		logger := log.With(di.logger, "buildid", buildID, "path", objFile.Path)
+		di.ensureUploaded(ctx, objFile)
+	}
+}
 
-		if di.alreadyUploading(buildID) {
-			continue
-		}
-		di.markAsUploading(buildID)
+func (di *Manager) ensureUploaded(ctx context.Context, objFile *objectfile.MappedObjectFile) {
+	buildID := objFile.BuildID
+	path := objFile.Path
 
-		src := di.debugInfoSrcPath(ctx, buildID, objFile)
-		if src == "" {
-			continue
-		}
-		if exists := di.exists(ctx, buildID, src); exists {
-			continue
-		}
+	logger := log.With(di.logger, "buildid", buildID, "path", path)
 
-		buf := flexbuf.New()
-		if err := di.Extract(ctx, buf, src); err != nil {
-			level.Debug(di.logger).Log("msg", "failed to extract debug information", "err", err, "buildID", buildID, "path", src)
-			continue
-		}
+	// removing the buildID from the cache to ensure a re-upload at the next interation.
+	defer di.removeAsUploading(buildID)
 
-		buf.SeekStart()
-		if err := validate(buf); err != nil {
-			level.Debug(logger).Log("msg", "failed to validate debug information", "err", err, "buildID", buildID, "path", src)
-			continue
-		}
+	if di.alreadyUploading(buildID) {
+		return
```

Comment:
Above we defer `removeAsUploading` in this case we will mark this build id as finished uploading prematurely.

Reply:
I think defer should move after this condition.

Reply:
Thanks for the review ! I'm actually wondering if the condition makes sense anymore.

When a new buildid is discovered, we will go through the entire upload path, the buildid will be added to the cache and due to the defer removed immediately. During the next iteration, `di.alreadyUploading` will be false and will always be false now. Right ? Or am I missing something ?

Reply:
This process is async that's the reason we added this cache-based lock in the first place (burrow cache is thread-safe). So we want to prevent re-uploading if there's already an upload in progress. So we should only remove the uploading state if we marked that build id as uploading.

Reply:
Ha yes sorry just saw we are calling this function from a go routine ! Thanks :)